File: tests/units/surface-jacobian-det/surface_jacobian_det.m
## Copyright (C) 2021-2023 Jihuan Tian <jihuan_tian@hotmail.com>
##
## This file is part of the HierBEM library.
##
## HierBEM is free software: you can use it, redistribute it and/or modify it
## under the terms of the GNU Lesser General Public License as published by the
## Free Software Foundation, either version 3 of the License, or (at your
## option) any later version. The full text of the license can be found in the
## file LICENSE at the top level directory of HierBEM.

clear all;
ConfigGraphicsToolkit;

addpath("/home/jihuan/Projects/git/octbem");
addpath("/home/jihuan/Projects/git/octbem/gmsh_io");

pkg load fpl;
pkg load msh;

mesh_filename = "./surface-jacobian-det-square.msh";
[path_name, file_name, file_ext] = fileparts(mesh_filename);
problem_domain_mesh = ReadGmshQuads(mesh_filename);

## ## Plot the mesh.
## figure;
## quadplot3d(problem_domain_mesh.mesh_cells, problem_domain_mesh.mesh_nodes, "color", "r", "linestyle", "-", "marker", ".");
## hold on;

## #############################################################################
## Generate the shape function space for describing the cell geometry, the order
## of which is determined by the mesher.
## #############################################################################
## At present, assume all cells in the mesh have the same geometric
## order, which is only 1st order at the moment.
shape_function_space_order = problem_domain_mesh.cell_geom_orders(2);
## Generate the series of shape functions dependent on the first two components
## of the area coordinate, while the last component is not independent.
shape_function_space = LagrangeBasisOn3DQuad(shape_function_space_order);
## Generate the local coordinates for the support nodes of shape functions.
shape_function_support_nodes = AreaCoordsOnQuad(shape_function_space_order);
## Total number of basis functions in the shape function space.
number_of_bases_in_shape_function_space = length(shape_function_space);

surface_jacobian_det_vector = zeros(problem_domain_mesh.number_of_cells, 1);
unit_cell_center = [0.5, 0.5];
for e = 1:problem_domain_mesh.number_of_cells
  J_functor = @(area_coord) GlobalSurfaceMetricOn3DQuad(area_coord, problem_domain_mesh.mesh_nodes(problem_domain_mesh.mesh_cells(e, :), :));
  surface_jacobian_det_vector(e) = J_functor(unit_cell_center);
  printf("%.5e\n", surface_jacobian_det_vector(e));
endfor

## Add a trailing newline to be consistent with the output of deal.ii.
printf("\n");

File: tests/units/surface-jacobian-det/surface-jacobian-det.cc
// Copyright (C) 2021-2025 Jihuan Tian <jihuan_tian@hotmail.com>
//
// This file is part of the HierBEM library.
//
// HierBEM is free software: you can use it, redistribute it and/or modify it
// under the terms of the GNU Lesser General Public License as published by the
// Free Software Foundation, either version 3 of the License, or (at your
// option) any later version. The full text of the license can be found in the
// file LICENSE at the top level directory of HierBEM.

// File: surface-jacobian-det.cc
// Description: This test calculates the surface Jacobian determinant for
// quadrangular cell. Author: Jihuan Tian Date: 2020-11-04 Copyright (C) 2020
// Jihuan Tian <jihuan_tian@hotmail.com>

#include <deal.II/base/logstream.h>

#include <deal.II/fe/fe_q.h>
#include <deal.II/fe/mapping_q.h>

#include <deal.II/grid/grid_generator.h>
#include <deal.II/grid/grid_in.h>
#include <deal.II/grid/grid_out.h>

#include <deal.II/lac/full_matrix.h>
#include <deal.II/lac/full_matrix.templates.h>

#include <fstream>
#include <iostream>

#include "laplace/laplace_bem.h"

int
main()
{
  deallog.pop();
  deallog.depth_console(2);

  const unsigned int dim      = 2;
  const unsigned int spacedim = 3;

  // Generate a sphere grid.
  Triangulation<dim, spacedim> triangulation;
  GridGenerator::hyper_sphere(triangulation);
  triangulation.refine_global(3);
  GridOut       gridout;
  std::ofstream mesh_file("./surface-jacobian-det-sphere.msh");
  gridout.write_msh(triangulation, mesh_file);

  //  // Generate a square grid.
  //  Triangulation<dim, spacedim> triangulation;
  //  GridGenerator::hyper_cube(triangulation);
  //  triangulation.refine_global(3);
  //  GridOut gridout;
  //  std::ofstream mesh_file("./surface-jacobian-det-square.msh");
  //  gridout.write_msh(triangulation, mesh_file);

  const unsigned int      n_active_cells = triangulation.n_active_cells();
  Vector<double>          cell_surface_jacobian_det(n_active_cells);
  const unsigned int      fe_degree = 2;
  FE_Q<dim, spacedim>     fe(fe_degree);
  MappingQ<dim, spacedim> mapping(fe_degree);

  // Iterative over each cell and evaluate the surface Jacobian determinant at a
  // specified unit cell point.
  Point<2> unit_cell_center(0.5, 0.5);

  unsigned int i = 0;
  for (const auto cell : triangulation.active_cell_iterators())
    {
      std::vector<Point<spacedim>> real_support_points(
        HierBEM::support_points_in_real_cell(
          cell, fe, mapping, fe.get_poly_space_numbering_inverse()));

      cell_surface_jacobian_det(i) =
        HierBEM::surface_jacobian_det(fe,
                                      real_support_points,
                                      unit_cell_center);

      i++;
    }

  cell_surface_jacobian_det.print(deallog.get_console(), 5, true, false);

  return 0;
}
